src/index.ts

```diff
@@ -1,32 +1,80 @@
 import { sign } from "./signature";
-import { setData } from "./privacy";
+import { SchematisedDocument, SignedDocument } from "./@types/document";
 import { saltData } from "./privacy/salt";
 import * as utils from "./utils";
+import openAttestationSchema from "./schema/2.0/schema.json";
+import { setData } from "./privacy";
+import * as v1 from "./__generated__/schemaV1";
+import * as v2 from "./__generated__/schemaV2";
+
+const createDocument = (data: any, schemaId?: string) => {
+  const version = openAttestationSchema.$id.split("/")?.[1];
+  if (!version) {
+    throw new Error(`Unable to detect the version of the current open-attestation schema`);
+  }
+  const documentSchema: SchematisedDocument = {
+    // throw error
+    version,
+    data: null
+  };
+  if (schemaId) {
+    documentSchema.schema = schemaId;
+  }
+  return setData(documentSchema, saltData(data));
+};
 
-const createDocument = (data: any, schema: Schema) => {
-  const document = setData({ schema: schema.$id, data: null }, saltData(data));
-  const valid = validate(document, schema);
-  if (valid) {
-    return document;
+class SchemaValidationError extends Error {
+  constructor(message: string, public validationErrors: Ajv.ErrorObject[], public document: any) {
+    super(message);
   }
-  throw new Error(`Invalid document:${JSON.stringify(data, null, 2)}`);
+}
+const isSchemaValidationError = (error: any): error is SchemaValidationError => {
+  return !!error.validationErrors;
 };
 
-export const issueDocument = (data: any, schema: Schema) => {
-  const document = createDocument(data, schema);
+interface IssueDocumentOption {
+  externalSchemaId?: string;
+}
+export const issueDocument = (
+  data: unknown,
+  options?: IssueDocumentOption
+): SignedDocument<v2.OpenAttestationDocument> => {
+  const document: SchematisedDocument = createDocument(data, options?.externalSchemaId);
+  const errors = validate(document, getSchema("open-attestation/2.0"));
+  if (errors.length > 0) {
+    throw new SchemaValidationError("Invalid document", errors, document);
+  }
   return sign(document, [digestDocument(document)]);
 };
 
-export const issueDocuments = (dataArray: any[], schema: Schema) => {
-  const documents = dataArray.map(data => createDocument(data, schema));
+export const issueDocuments = (
+  dataArray: unknown[],
+  options?: IssueDocumentOption
+): SignedDocument<v2.OpenAttestationDocument>[] => {
+  const documents = dataArray.map(data => createDocument(data, options?.externalSchemaId));
+  documents.forEach(document => {
+    const errors = validate(document, getSchema("open-attestation/2.0"));
+    if (errors.length > 0) {
+      throw new SchemaValidationError("Invalid document", errors, document);
+    }
+  });
+
   const batchHashes = documents.map(digestDocument);
   return documents.map(doc => sign(doc, batchHashes));
 };
 
+export const validateSchema = (document: SignedDocument): boolean => {
```

Comment:
This does not work for existing OAv1 document as there is no `version: "1.0"` in the files.